### ISSUE TITLE
after building latest pull, symbol lookup error 

### ISSUE THREAD

**Issue by moshix** [SDL-Hercules-390/hyperion#127]
I am building the latest pull on Ubuntu 18.04 and build goes without any isues, but when I start hercules, I get the following:

~/hyperion$ hercules
hercules: symbol lookup error: hercules: undefined symbol: nix_set_thread_name

Not sure what I can do to prevent this error. The build from a few days ago works fine. 

thanks

Moshix

**Comment by ivan-w** [comment]
You probably need to indicate *HOW* you are building hercules ;)

**Comment by moshix** [comment]
Fair enough, Ivan. 

I git clone this repository. Then I apt install all the pre-requisites such as the zlib and bzlib. 

Then, I do:

./autogen.sh
./configure
make
make install

This all goes well without any errors, but when I then launch hercules I get the above error of symbol lookup. 

Many thanks!

M

**Comment by ivan-w** [comment]
I assume that after "make install" you go to a directory where your hercules.cnf lives and your PATH contains /usr/local/hercules/bin (since you didn't specify any argument to ./configure)...

**Comment by ivan-w** [comment]
On a side note, something that *always* works for me is to have separate source, build and run directories..

The main advantages are
- the builds are always clean and pristine
- You can have a indefinite number of different builds living on the same system (it's just a matter of setting the PATH) - VERY practical when doing "git bisect" and you need to retrace your steps...
On a modern x86 system (say Core i7), building hercules from scratch with -O3 and -j <number of cores*threads> usually takes less than 2 minutes..
- You can also experiment with various compilers without having to rebuild after every test

Here is my build script : 

```
#!/bin/bash

if [[ $# -lt 2 ]]
then
        echo "Missing parameter"
        exit 1
fi
p1=$(basename $1)
p2=$(basename $2)
src=$(realpath ../src/$p1)
bld=$(realpath $p2)
if [[ $# -lt 3 ]]
then
        run=$(realpath ../run/$p2)
else
        run=$3
fi

echo "Using Source $src"
echo "Using Build $bld"
echo "Using Run $run"

rm -rf $bld
mkdir -p $bld
cd $bld
export MAX_CPU_ENGINES=32
$src/configure --prefix=$run --enable-optimization="-O3 -march=native -fomit-frame-pointer"
make -j
make install

```

**Comment by moshix** [comment]
yes, I do go to the directories, where my .cnf is. 

I have built hyperion hundreds of times, and it's always worked. What I am saying is that on ubuntu 18.04 I have problems building it. On all other systems I have, it builds and works fine. 

And yes, I similar system to yours to build (I was simplifyingto eliminate my build system as the cause). 

thanks

**Comment by moshix** [comment]
oh and by the way, I don't turn on O3 optimization because I found that with z/VM 6.x and more than 5 CPUs, O3 reliably gets z/VM to dump and stop all the CPUs, when on a NUMA system (like my trusty HP 580 G8). On non-NUMA machines O3 works perfectly well. 

I guess there are still some race conditions in the code. 

thanks a lot Ivan!

**Comment by Fish-Git** [comment]
> What I am saying is that on Ubuntu 18.04 I have problems building it. _On all other systems I have, it builds and works fine._

Is your Ubuntu 18.04 system using the same compiler version as your other systems? (The hercules `version` command displays how Hercules was built, same as upon startup.)

> I found that with z/VM 6.x and more than 5 CPUs, O3 reliably gets z/VM to dump and stop all the CPUs, _when on a NUMA system_ (like my trusty HP 580 G8). _On non-NUMA machines O3 works perfectly well._

Interesting!

I would be _very_ interested in learning whether or not the same problem occurs on Windows on the same NUMA hardware. This is beginning to sound like a Linux (Ubuntu) bug related to NUMA hardware and not necessarily a Hercules bug.

I will understand if you do not wish to take the time to do the following, but I feel I must ask anyway:

Would it be possible for you to install Windows on your NUMA HP 580 G8 system (inside a virtual machine) and run the same test? I would be _very_ interested to learn whether running z/VM 6.x under Hercules on Windows with more than 5 virtual CPUs causes z/VM to crash or not.

Thanks!

**Comment by Fish-Git** [comment]
> Regarding NUMA:
> 
> Yes, I also thought of trying out Windows on my HP 580 (with four populated sockets and about 40 cores in total) and - importantly - 512GB of RAM over the two NUMA zones of this machine.

WOW! :astonished:

That is _impressive_ hardware to be running Hercules on! How many virtual CPUs do you define to Hercules (MAXCPU/NUMCPU) and what MIPS rate do you typically see?

> No ubuntu 18.04 has a different version of the compiler:
> gcc --version
> gcc (Ubuntu 7.2.0-18ubuntu2) 7.2.0
> 
> Which is a much newer version than what I have on Ubuntu 16.04 or even 14.04 (which is what I mostly use for hercules stuff).

Sounds like a compiler bug to me.

Does using clang instead of gcc make any difference?

**Comment by moshix** [comment]
I run a lot of VMs on that machine, not all hercules related, so the memory comes in handy. However, the issue with NUMA is that you don't really know when you do an malloc() if you get all memory in the same NUMA zone. the underlying OS (in this case Vmware ESXi) should try to do all mallocs intelligently, but there is no guarantee. 

I typically allocate 6-8 CPUs to Hercules. I don't even try to do more, because I am sure NUMA actually hurts performance beyond that. I do give sometimes 64GB or RAM to hercules 64bit OSs. z/VM up to 6.3 don't react well to this much RAM on Hercules. On the real machines of course there is no problem.

**Comment by Fish-Git** [comment]
> I did try with Windows 10 and same problem again as Linux, so I can rule out the OS.

Hmmm... Okay. Then it's definitely a bug in Hercules. The z/VM crashing that is! _Not_ the problem reported in this issue!

We seem to be cross-talking about two different problems in two different GitHub Issues: one is the z/VM crashing (that's issue #126), and the other is "nix_set_thread_name" undefined symbol at startup (that's _**this issue**_: #127).

**Comment by moshix** [comment]
Yes, I also thought of that. So let me open a new bug for the NUMA issue then.

**Comment by moshix** [comment]
moshix@xps15:~$ uname -a
Linux xps15 4.15.0-32-generic #35-Ubuntu SMP Fri Aug 10 17:58:07 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
moshix@xps15:~$ nm --dynamic /usr/local/lib/hercules/hutil.so | grep nix_set_thread_name
nm: '/usr/local/lib/hercules/hutil.so': No such file
moshix@xps15:~$ which hercules
/usr/local/bin/hercules
moshix@xps15:~$ nm --dynamic /usr/local/bin/hercules | grep nix_set_thread_name
                 U nix_set_thread_name

voila'
thanks

**Comment by moshix** [comment]
Managed to make SDL-hyperion latest work fine by turning on optimization and recompiling. I can't fathom how optimization has anything to do with it. 

On my laptop I usually turn off optimization since it's not performance that I am after on a laptop.

**Comment by Fish-Git** [comment]
> nm: '/usr/local/lib/hercules/hutil.so': No such file

Damn! It's the nm for hutil that I was mostly interested.&nbsp;`:(`

I'm trying to verify that the `nix_set_thread_name` symbol _is_ actually defined in the hutil shared library.&nbsp;&nbsp;&nbsp;_(If it's not, then something is very wrong somewhere!)_

I apologize for providing incorrect instructions to you. I'm not a Linux person and should have tested the commands on my own Linux system before telling you to do them, but I didn't. My apologies.

It appears the "hutil" shared library is _actually_ called `libhercu.so` and, if you did a default install of Hercules, _should_ exist in your `/usr/local/lib` directory.&nbsp;&nbsp;&nbsp;_(Please verify!)_

Thus, the _correct_ command (unless I screwed up again somewhere!) _should_ actually be:
```
    nm --dynamic /usr/local/lib/libhercu.so | grep nix_set_thread_name
```
Please try that instead and report back with the results.

Thanks!
